### ISSUE TITLE
Generic type is not expanded  when typechecking equality check

### ISSUE THREAD

**Issue by fictitious** [Microsoft/TypeScript#26758]
**TypeScript Version:**   3.1.0-dev.20180829

**Search Terms:** generics expansion union equality typecheck

**Code**

```ts
type A = {type :"a"}
type B = {type :"b"}

type P<T extends A | B> = T["type"];

class C<T extends A | B> {

    p: P<T> = 'x'; //error: Type '"x"' is not assignable to type '"a" | "b"'
    // so when typechecking this assignment, it's known that 
    // P<T> actually is "a" | "b"

    f(p: P<T>) {
        if (p === 'x') { // why no error here
        // when typechecking this equality check, P<T> is not expanded ?   
        }
    }

    // expected error, for illustration
    f0(p: 'a' | 'b') {
        if (p === 'x') { // error: This condition will always return 'false' 
                         // since the types '"a" | "b"' and '"x"' have no overlap.

        }
    }

}
```

**Expected behavior:**

"This condition will always return 'false'" is reported when doing equality check between a variable declared with `P<T>` type and "x" literal

**Actual behavior:**

No error is reported for that particular equality check

**Playground Link:**  [link](https://www.typescriptlang.org/play/#src=type%20A%20%3D%20%7Btype%20%3A%22a%22%7D%0D%0Atype%20B%20%3D%20%7Btype%20%3A%22b%22%7D%0D%0A%0D%0Atype%20P%3CT%20extends%20A%20%7C%20B%3E%20%3D%20T%5B%22type%22%5D%3B%0D%0A%0D%0Aclass%20C%3CT%20extends%20A%20%7C%20B%3E%20%7B%0D%0A%0D%0A%20%20%20%20p%3A%20P%3CT%3E%20%3D%20'x'%3B%20%2F%2Ferror%3A%20Type%20'%22x%22'%20is%20not%20assignable%20to%20type%20'%22a%22%20%7C%20%22b%22'%0D%0A%20%20%20%20%2F%2F%20so%20when%20typechecking%20this%20assignment%2C%20it's%20known%20that%20%0D%0A%20%20%20%20%2F%2F%20P%3CT%3E%20actually%20is%20%22a%22%20%7C%20%22b%22%0D%0A%0D%0A%20%20%20%20f(p%3A%20P%3CT%3E)%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(p%20%3D%3D%3D%20'x')%20%7B%20%2F%2F%20why%20no%20error%20here%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20when%20typechecking%20this%20equality%20check%2C%20P%3CT%3E%20is%20not%20expanded%20%3F%20%20%20%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20%2F%2F%20expected%20error%2C%20for%20illustration%0D%0A%20%20%20%20f0(p%3A%20'a'%20%7C%20'b')%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(p%20%3D%3D%3D%20'x')%20%7B%20%2F%2F%20error%3A%20This%20condition%20will%20always%20return%20'false'%20%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20since%20the%20types%20'%22a%22%20%7C%20%22b%22'%20and%20'%22x%22'%20have%20no%20overlap.%0D%0A%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%7D%0D%0A)

**Related Issues:** did not find anything

**Comment by mattmccutchen** [comment]
Analysis (originally posted [here](https://stackoverflow.com/questions/52087511)): The rule for `a === b` is normally that either the type of `a` has to be comparable to the type of `b` or vice versa, but there is a special case: if one type is a union of literals and the other type is not, then the type that is a union of literals is changed to the underlying primitive type (here `string`) for the purpose of the check.  This special case is firing undesirably on the first `p === 'x'` where one side is a union of literals and the other side is a type parameter that is not itself a union of literals but is _constrained_ by a union of literals.

Additional search terms: comparable constrained constraint union literal

**Comment by mattmccutchen** [comment]
Fix is in #27588.
